x-pack/legacy/plugins/maps/public/angular/map_controller.js

```diff
@@ -124,6 +129,37 @@ app.controller('GisMapController', ($scope, $route, config, kbnUrl, localStorage
     store.dispatch(setRefreshConfig($scope.refreshConfig));
   };
 
+
+  function hasUnsavedChanges() {
+
+    const state = store.getState();
+    const layerList = getLayerListRaw(state);
+    const layerListConfigOnly = copyPersistentState(layerList);
+
+    const savedLayerList  = savedMap.getLayerListJSON();
+    const oldConfig = savedLayerList ? savedLayerList : initialLayerListConfig;
+
+    return !_.isEqual(layerListConfigOnly, oldConfig);
+  }
+
+  function isOnMapNow() {
+    return window.location.hash.startsWith(`#/${MAP_SAVED_OBJECT_TYPE}`);
+  }
+
+  function beforeUnload(event) {
+    if (!isOnMapNow()) {
+      return;
+    }
+
+    const hasChanged = hasUnsavedChanges();
+    if (hasChanged) {
+      event.preventDefault();
+      event.returnValue = 'foobar';//this is required for Chrome
+    }
+  }
+  window.removeEventListener('beforeunload', beforeUnload);
```

Comment:
Does this also need to be called in `$destroy`?

Reply:
Just did a test and multiple `beforeUnload` handlers are getting registered. Toggle between maps listing page and creating a map a few times. Then put a break point in beforeUnload function, add a layer and navigate away from Maps applicaiton. `beforeUnload` is called for each instance that MapController was created. Putting `window.removeEventListener('beforeunload', beforeUnload);` in $destroy handler resolved the problem.